src/lib/support/jsontlv/TlvToJson.cpp

```diff
@@ -0,0 +1,356 @@
+/*
+ *
+ *    Copyright (c) 2023 Project CHIP Authors
+ *
+ *    Licensed under the Apache License, Version 2.0 (the "License");
+ *    you may not use this file except in compliance with the License.
+ *    You may obtain a copy of the License at
+ *
+ *        http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+
+#include "lib/support/CHIPMemString.h"
+#include "lib/support/ScopedBuffer.h"
+#include <json/json.h>
+#include <lib/core/DataModelTypes.h>
+#include <lib/support/Base64.h>
+#include <lib/support/SafeInt.h>
+#include <lib/support/jsontlv/ElementTypes.h>
+#include <lib/support/jsontlv/TlvToJson.h>
+
+namespace chip {
+
+namespace {
+
+constexpr uint32_t kTemporaryImplicitProfileId = 0x1122;
+
+/// RAII to switch the implicit profie id for a reader
+class ImplicitProfileIdChange
+{
+public:
+    ImplicitProfileIdChange(TLV::TLVReader & reader, uint32_t id) : mReader(reader), mOldImplicitProfileId(reader.ImplicitProfileId)
+    {
+        reader.ImplicitProfileId = id;
+    }
+    ~ImplicitProfileIdChange() { mReader.ImplicitProfileId = mOldImplicitProfileId; }
+
+private:
+    TLV::TLVReader & mReader;
+    uint32_t mOldImplicitProfileId;
+};
+
+const char * GetJsonElementStrFromType(const ElementTypeContext & ctx)
+{
+    switch (ctx.tlvType)
+    {
+    case TLV::kTLVType_UnsignedInteger:
+        return kElementTypeUInt;
+    case TLV::kTLVType_SignedInteger:
+        return kElementTypeInt;
+    case TLV::kTLVType_Boolean:
+        return kElementTypeBool;
+    case TLV::kTLVType_FloatingPointNumber:
+        return ctx.isDouble ? kElementTypeDouble : kElementTypeFloat;
+    case TLV::kTLVType_ByteString:
+        return kElementTypeBytes;
+    case TLV::kTLVType_UTF8String:
+        return kElementTypeString;
+    case TLV::kTLVType_Null:
+        return kElementTypeNull;
+    case TLV::kTLVType_Structure:
+        return kElementTypeStruct;
+    case TLV::kTLVType_Array:
+        return kElementTypeArray;
+    default:
+        return kElementTypeEmpty;
+    }
+};
+
+/*
+ * Encapsulates the element information required to constract Json element name string in a Json object.
```

Comment:
The "Json" spelling in all the comments here is odd.  It should all be "JSON" or "json"....